rfc.md

```diff
@@ -0,0 +1,167 @@
+
+### Non-Goals
+
+- Rewriting the existing bootloader backends to actually fill out the feature matrix. The goal of this RFC is to make the feature development *easier*, not actually do it.
+- Supporting SecureBoot. The authors of this RFC have done work in this regard, but this RFC is not about SecureBoot.
+
+# Proposed Solution
+
+- Each NixOS generation will have a bootspec (a JSON document) at `$out/boot.v1.json` containing all of the boot properties for that generation. NixOS’s bootloader backends will read these files as inputs to the bootloader installation phase.
+    - The bootloader installation phase is relatively unchanged from the way it is now. The bootloader backend will have an executable that is run against a collection of generations, and the backend is any of the currently supported backends plus an “external” backend which the user can define.
+- The bootloader backends will avoid reading data from the other files and directories when possible, preferring the information in the bootspec.
+- A bootspec synthesizing tool will be used to synthesize a bootspec for generations which don’t already have one. This tool will be shared across all of the bootloader backends, helping produce more uniform behavior.
+- Existing bootloader backends will be updated to read properties from the bootspec, removing most if not all of their filesystem-spelunking code.
+
+### Bootspec Format v1
+
+Using the following JSON:
+
+```yaml
+{
+  # Version of the specification used in the document
+  "schemaVersion": 1,
+
+  # Path to the stage-2 init, executed by the initrd
+  "init": "/nix/store/xxx-nixos-system-xxx/init",
+
+  # Path to the initrd
+  "initrd": "/nix/store/xxx-initrd-linux/initrd",
+
+  # Optional path to a tool to dynamically add secrets to an initrd
+  "initrdSecrets": "/nix/store/xxx-append-secrets/bin/append-initrd-secrets",
+
+  # Path to the kernel image
+  "kernel": "/nix/store/xxx-linux/bzImage",
+
+  # Kernel commandline options
+  "kernelParams": [
+    "amd_iommu=on",
+    "amd_iommu=pt",
+    "iommu=pt",
+    "kvm.ignore_msrs=1",
+    "kvm.report_ignored_msrs=0",
+    "udev.log_priority=3",
+    "systemd.unified_cgroup_hierarchy=1",
+    "loglevel=4"
+  ],
+
+  # The label of the system. It should contain the operating system, kernel version,
+  # and other user-relevant information to identify the system. This corresponds
+  # loosely to `config.system.nixos.label`.
+  "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",
+
+  # Top level path of the closure, in case some spelunking is required
+  "toplevel": "/nix/store/xxx-nixos-system-xxx",
+
+  "specialisation": {
+    # <name> corresponds to <name> in specialisation.<name>.configuration.
+    # Note: a specialisation's bootspec document should not contain any specialisations.
+    "<name>": {
+      # bootspec is the path to a bootspec document
+      "bootspec": "/path/to/a/bootspec.v1.json",
+    }
+  }
+}
```

Comment:
nit: Use JSON5 for example json document

🎉 trailing commas! 🎉 comments! 🎉 

```suggestion
```json5
{
  // Version of the specification used in the document
  "schemaVersion": 1,

  // Path to the stage-2 init, executed by the initrd
  "init": "/nix/store/xxx-nixos-system-xxx/init",

  // Path to the initrd
  "initrd": "/nix/store/xxx-initrd-linux/initrd",

  // Optional path to a tool to dynamically add secrets to an initrd
  "initrdSecrets": "/nix/store/xxx-append-secrets/bin/append-initrd-secrets",

  // Path to the kernel image
  "kernel": "/nix/store/xxx-linux/bzImage",

  // Kernel commandline options
  "kernelParams": [
    "amd_iommu=on",
    "amd_iommu=pt",
    "iommu=pt",
    "kvm.ignore_msrs=1",
    "kvm.report_ignored_msrs=0",
    "udev.log_priority=3",
    "systemd.unified_cgroup_hierarchy=1",
    "loglevel=4",
  ],

  // The label of the system. It should contain the operating system, kernel version,
  // and other user-relevant information to identify the system. This corresponds
  // loosely to `config.system.nixos.label`.
  "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",

  // Top level path of the closure, in case some spelunking is required
  "toplevel": "/nix/store/xxx-nixos-system-xxx",

  "specialisation": {
    // <name> corresponds to <name> in specialisation.<name>.configuration.
    // Note: a specialisation's bootspec document should not contain any specialisations.
    "<name>": {
      // bootspec is the path to a bootspec document
      "bootspec": "/path/to/a/bootspec.v1.json",
    },
  },
}
```

Reply:
:+1: applied and pushed, thank you!